### ISSUE TITLE
There is no tracker/neuralnet in the list.

### ISSUE THREAD

**Issue by derrikgort** [opentrack/opentrack#1394]
I don't see no tracker/neuralnet, though the opentrack was run under the admin root. This issue is on 2021.2.0, 2021.1.3.
![snapx4653](https://user-images.githubusercontent.com/40070032/146678238-3bbd8d9a-c1b5-4bfc-b925-87fc362f7767.jpg)

**Comment by sthalik** [comment]
Probably missing Visual Studio 2015 Redistributable.

**Comment by derrikgort** [comment]
All visuals and frameworks are installed.

**Comment by sthalik** [comment]
`opentrack-tracker-neuralnet.dll` exists in the modules directory, no?

**Comment by derrikgort** [comment]
Of course it exists
![snapx4654](https://user-images.githubusercontent.com/40070032/146679835-5092684f-b4d4-44af-8b77-b5bd7ccaf69b.jpg)
.

**Comment by sthalik** [comment]
Then drag opentrack.exe onto a Windows Console window, press Return and see why it fails to load neuralnet.

**Comment by derrikgort** [comment]
![snapx4655](https://user-images.githubusercontent.com/40070032/146680184-9956f50e-3faa-4fe9-96bc-7df2bcaf18da.jpg)

**Comment by sthalik** [comment]
Then run Dependency Walker. Just need to set up PATH first. Here's how.

![image](https://user-images.githubusercontent.com/1896811/146680719-7cc774e2-d42a-42be-9b93-0d98a9ac2198.png)
![image](https://user-images.githubusercontent.com/1896811/146680757-d8a4aa47-a0c4-4389-9142-5681c69babc5.png)

On my end no libraries are missing, so the thing works:

![image](https://user-images.githubusercontent.com/1896811/146680785-5846e34d-c830-429e-b26f-288d62c0ef72.png)

But really I should be making `-MT` builds for the ONNX runtime.

**Comment by derrikgort** [comment]
Sorry, I'm not a programmer, I can run the opentrack in cmd.exe but then I just don't get what I should set up a path for and what is dependency walker, and what I have to do with them. I just report an issue I got as a user.

**Comment by GO63-samara** [comment]
I see that you don't have `neuralnet` and `kinectface` trackers on your list.
Do you have Windows 7? These trackers do not work in Windows 7. You must be using Windows 10.
When I launch Opentrack-21.2.0 from cmd in Windows 7, two libraries are not loaded:
```
D:\TOOLS_WIN\AVIA\OpenTrack\opentrack-2021.2.0>
DEBUG [F:\dev\opentrack\api/plugin-support.hpp:107]: library "kinect-face" faile
d: "Cannot load library D:\\TOOLS_WIN\\AVIA\\OpenTrack\\opentrack-2021.2.0\\modu
les\\\\opentrack-tracker-kinect-face.dll: DEBUG [F:\dev\opentrack\api/plugin-sup
port.hpp:107]: library "neuralnet" failed: "Cannot load library D:\\TOOLS_WIN\\A
VIA\\OpenTrack\\opentrack-2021.2.0\\modules\\\\opentrack-tracker-neuralnet.dll:
```
But in Windows 10 is all fine. The list contains all 17 trackers.

Try this:
Windows 7 runs `neuralnet-0.2` from here:
https://github.com/DaWelter/opentrack/releases
Direct link:
https://github.com/DaWelter/opentrack/releases/download/tracker-neuralnet-0.2/opentrack-portable-tracker-neuralnet-v0.2.7z

**Comment by derrikgort** [comment]
> I see that you don't have `neuralnet` and `kinectface` trackers on your list. Do you have Windows 7? These trackers do not work in Windows 7. You must be using Windows 10. When I launch Opentrack-21.2.0 from cmd in Windows 7, two libraries are not loaded:
> 
> ```
> D:\TOOLS_WIN\AVIA\OpenTrack\opentrack-2021.2.0>
> DEBUG [F:\dev\opentrack\api/plugin-support.hpp:107]: library "kinect-face" faile
> d: "Cannot load library D:\\TOOLS_WIN\\AVIA\\OpenTrack\\opentrack-2021.2.0\\modu
> les\\\\opentrack-tracker-kinect-face.dll: DEBUG [F:\dev\opentrack\api/plugin-sup
> port.hpp:107]: library "neuralnet" failed: "Cannot load library D:\\TOOLS_WIN\\A
> VIA\\OpenTrack\\opentrack-2021.2.0\\modules\\\\opentrack-tracker-neuralnet.dll:
> ```
> 
> But in Windows 10 is all fine. The list contains all 17 trackers.
> 
> Try this: Windows 7 runs `neuralnet-0.2` from here: https://github.com/DaWelter/opentrack/releases Direct link: https://github.com/DaWelter/opentrack/releases/download/tracker-neuralnet-0.2/opentrack-portable-tracker-neuralnet-v0.2.7z

Yes, you are right, I'm on win7. I've downloaded the build you offered and it works fine...Now all I have to test it in work. I'm using aruco, it works ok, I wonder if the neuralnet will be better...

**Comment by GO63-samara** [comment]
@sthalik
I see in your screenshot there is no `kinect-face tracker`, but there is an unknown `TrackHat Point Tracker`. This is fine?

**Comment by derrikgort** [comment]
Can the neuralnet be working on the next versions of opentrack under win7? Thanks for advance.

**Comment by sthalik** [comment]
@GO63-samara 
> I see in your screenshot there is no kinect-face tracker, but there is an unknown TrackHat Point Tracker. This is fine?

If your `neuralnet` build works, is this because you're building with `-MT`, or for some other reason?

The TrackHat sensor support hasn't been released yet. Kinect is missing because I only have parts of the SDK on the build machine.

**Comment by sthalik** [comment]
@derrikgort please see if this build works. It was built with the official release version of onnxruntime. https://www.dropbox.com/s/hllwehp4j8q0cpi/opentrack-master-20211219_01.7z

**Comment by derrikgort** [comment]
No, it is still absent... Track hat point tracker appeared ( It wasn't last time too), but there is no neuralnet again..
![snapx4656](https://user-images.githubusercontent.com/40070032/146690367-5c7ef29b-bc0b-42d0-976c-36e2afb362e8.jpg)
.

**Comment by GO63-samara** [comment]
> If your neuralnet build works, is this because you're building ONNX with -MT, or for some other reason?

This is not my build, this is the build from @DaWelter.

I didn't manage to build `neuralnet` myself. I did not use the sources, but the `onnxruntime-1.8.1` release. I have configured three lines in `sdk-paths`:

```
setq(ONNXRuntime_INCLUDE_DIR "onnxruntime-1.8.1/include")
setq(ONNXRuntime_LIBRARY "onnxruntime-1.8.1/lib/onnxruntime.lib")
setq(ONNXRuntime_RUNTIME "onnxruntime-1.8.1/lib/onnxruntime.dll")
```
but apparently this is not enough. I got build errors.

**Comment by GO63-samara** [comment]
When launched in Windows7 `OT-20211219_01` `neuralnet` does not appear:

```
D:\TOOLS_WIN\AVIA\OpenTrack\OT-20211219_01>
DEBUG [F:\dev\opentrack\api/plugin-support.hpp:106]: library "kinect-face" faile
d: "Cannot load library D:\\TOOLS_WIN\\AVIA\\OpenTrack\\OT-20211219_01\\modules\
\\\opentrack-tracker-kinect-face.dll: DEBUG [F:\dev\opentrack\api/plugin-support
.hpp:106]: library "neuralnet" failed: "Cannot load library D:\\TOOLS_WIN\\AVIA\
\OpenTrack\\OT-20211219_01\\modules\\\\opentrack-tracker-neuralnet.dll:
```

**Comment by sthalik** [comment]
See if this build works: https://www.dropbox.com/s/7m27azwabkc9ik2/opentrack-master-20211220_01.7z

**Comment by DaWelter** [comment]
Hi guys
thank you for the support.

> but apparently this is not enough. I got build errors.

When I started using the official dll's, setting the ONNX dir like so `set(ONNXRuntime_DIR "D:/dev/onnxruntime-official-1.9-x86-release/" CACHE PATH "" FORCE)` in my `sdk-paths-...@MSVC-windows.cmake` worked for me.

**Comment by derrikgort** [comment]
> See if this build works: https://www.dropbox.com/s/7m27azwabkc9ik2/opentrack-master-20211220_01.7z

Finally it works! Thanks. It works a bit strange, something wrong with centering, but perhaps it needs adjusting, anyway I have to test it later.

**Comment by GO63-samara** [comment]
Yes, in Windows 7, when launching OT-20191220_01, neuralnet appeared.
Maybe in a similar way, you can launch kinect-face in Windows 7?
```
D:\TOOLS_WIN\AVIA\OpenTrack\OT-20191220_01>
DEBUG [F:\dev\opentrack\api/plugin-support.hpp:106]: library "kinect-face" faile
d: "Cannot load library D:\\TOOLS_WIN\\AVIA\\OpenTrack\\OT-20191220_01\\modules\
\\\opentrack-tracker-kinect-face.dll:
```
Thanks!

**Comment by sthalik** [comment]
@GO63-samara,
> Maybe in a similar way, you can launch kinect-face in Windows 7?

Kinect SDK is binary with no source available. There's nothing that could be linked statically.

**Comment by sthalik** [comment]
It's no longer necessary to run `cmake .` after adding or removing a file, or switching branches.

```cmake
file(GLOB cxx CONFIGURE_DEPENDS "${dir}/*.cpp")
```

Now the build system does that by itself.

**Comment by GO63-samara** [comment]
> When I started using the official dll's, setting the ONNX dir like so set(ONNXRuntime_DIR "D:/dev/onnxruntime-official-1.9-x86-release/" CACHE PATH "" FORCE) in my sdk-paths-...@MSVC-windows.cmake worked for me.

@DaWelter
Yes, this line replaces three lines at once and does the same:
```
#setq(ONNXRuntime_INCLUDE_DIR "onnxruntime-1.8.1/include")
#setq(ONNXRuntime_LIBRARY "onnxruntime-1.8.1/lib/onnxruntime.lib")
#setq(ONNXRuntime_RUNTIME "onnxruntime-1.8.1/lib/onnxruntime.dll")

setq(ONNXRuntime_DIR "onnxruntime-1.8.1" CACHE PATH "" FORCE)
```
But I still get the error on the `ninja install` command:
```
ftnoir_tracker_neuralnet.cpp.obj : error LNK2001: unresolved external symbol _OrtGetApiBase@0
tracker-neuralnet\opentrack-tracker-neuralnet.dll : fatal error LNK1120: 1 unresolved externals
```
Any ideas?

**Comment by sthalik** [comment]
It works for me when using the toolchain file. Disabled all cmake options with `SHARED` in the name, except for `onnxruntime_BUILD_SHARED_LIB`.

Then,

```cmake
FLAGS_LD:STRING=-WX:NO dbghelp.lib # toolchain specific
ONNX_USE_MSVC_STATIC_RUNTIME:BOOL=TRUE
```

@GO63-samara, can you verify that there's `onnxruntime.lib` on the linker invocation command-line?

**Comment by GO63-samara** [comment]
@sthalik 
I tried to rebuild the `opentrack` today
I re-cloned `opentrack` and `opentrack-depends` 
I was unable to build the release.

Problem in `ninja install` command after connecting in `sdk-paths` path to OpenCV:

A lot of warning C4275 appeared like this:
```
Building CXX object filter-kalman\CMakeFiles\opentrack-filter-kalman.dir\moc_kalman.cpp.obj
D:\Dev\master\opentrack\api/plugin-api.hpp(94): warning C4275: non dll-interface struct 'TR' used as base for dll-interf
ace class 'Metadata'
D:\Dev\master\opentrack\api\../compat/tr.hpp(8): note: see declaration of 'TR'
D:\Dev\master\opentrack\api/plugin-api.hpp(94): note: see declaration of 'Metadata'
```

And the error appeared:
```
fatal error C1047: The object or library file 'D:\Dev\master\opentrack\..\..\opentrack-depends\libusb-msvc-x86\libusb-1.
0.lib' was created by a different version of the compiler than other objects like 'video-ps3eye\ps3eye-driver\CMakeFiles
\ps3eye-mode-test.dir\mode-test.cxx.obj'; rebuild all objects and libraries with the same compiler
LINK : fatal error LNK1257: code generation failed
[106/619] Linking CXX shared library compat\opentrack-compat.dll
ninja: build stopped: subcommand failed.
```

Maybe it's incompatible with **MSVC2019**?

**Comment by sthalik** [comment]
Rebuild libusb from version 1.0.23 as a static binary (after adjusting .vcxproj for -MT and -GL). Then replace `opentrack-depends/libusb-msvc-x86/libusb-1.0.lib`.

**Comment by GO63-samara** [comment]
I have no idea how to build libusb.
Instead, I took the previous version of libusb, before the update https://github.com/opentrack/opentrack-depends/commit/4dc314e16298f7243988a4b0e68b4c797c04d7d6

My main mistake was that I didn't rebuild OpenCV after the upgrade. Now I have rebuilt openCV again, with an updated `msvc.msvc` toolchain.
I also included the `WITH_ONNX ON` flag.

The `ninja install` command still has a lot of warnings like this:
```
[264/619] Building CXX object tracker-udp\CMakeFiles\opentrack-tracker-udp.dir\ftnoir_tracker_udp_dialog.cpp.obj
D:\Dev\master\opentrack\api/plugin-api.hpp(94): warning C4275: non dll-interface struct 'TR' used as base for dll-interf
ace class 'Metadata'
D:\Dev\master\opentrack\api\../compat/tr.hpp(8): note: see declaration of 'TR'
D:\Dev\master\opentrack\api/plugin-api.hpp(94): note: see declaration of 'Metadata'
```

The release now includes video trackers and even a neuralnet tracker. And it even works!

![image](https://user-images.githubusercontent.com/63638386/149021643-35e00a0f-3abf-42c9-a166-0ff2c2091160.png)

When I close the built Opentrack, I get a question in cmd:
```
D:\Dev\opentrack-build\install>
[ INFO:0] global C:\OpenCV\opencv-4.5.2\modules\core\src\parallel\registry_parallel.impl.hpp (96) cv::parallel::Parallel
BackendRegistry::ParallelBackendRegistry core(parallel): Enabled backends(0, sorted by priority): N/A
WARN QWinEventNotifier: no event dispatcher, application shutting down? Cannot deliver event.

D:\Dev\opentrack-build\install>

```
Thanks for the support.

I think you are better able to update your `libusb` version in `opentrack-depends`, right?

**Comment by sthalik** [comment]
> I have no idea how to build libusb.

```console
% git clone https://github.com/libusb/libusb.git
% cd libusb
% git checkout v1.0.23
Note: switching to 'v1.0.23'.
% start msvc/libusb_static_2017.vcxproj
```

Upgrade the SDK to latest. Switch to Release. Build.

Replace `libusb.h`, `libusb-1.0.lib` and `libusb-1.0.pdb` from the files that were created.

> I think you are better able to update your libusb version in opentrack-depends, right?

It's made for VS 2022.

**Comment by GO63-samara** [comment]
Thanks, this is a good guide.

I built `libusb v1.0.23` in MSVC2019.  /MT and /GL have already been set in `libusb Debug Properties`.

The `ninja install` command still has a lot of warnings like:
```
[449/627] Building CXX object spline\CMakeFiles\opentrack-spline.dir\axis-opts.cpp.obj
D:\Dev\master\opentrack\api/plugin-api.hpp(94): warning C4275: non dll-interface struct 'TR' used as base for dll-interf
ace class 'Metadata'
D:\Dev\master\opentrack\api\../compat/tr.hpp(8): note: see declaration of 'TR'
D:\Dev\master\opentrack\api/plugin-api.hpp(94): note: see declaration of 'Metadata'
```
But the question disappeared when exiting opentrack.

I think it would be more correct for me to install MSVC2022 and build OT in it.

**Comment by kanzaki-kunbr** [comment]
helo I'm having a problem in neuralnet it doesn't appear in opentrack
![unknown](https://user-images.githubusercontent.com/104150366/164477248-80292547-54ce-4d1d-adb4-56298a44d3e3.png)

**Comment by sthalik** [comment]
@kanzaki-kunbr Install the Visual Studio 2015 Redistributable. I'm going to fix this for future releases.

**Comment by kanzaki-kunbr** [comment]
@sthalik 
![image](https://user-images.githubusercontent.com/104150366/164481830-679cfa9f-b3c5-4e8e-9d37-15257cdccccf.png)

**Comment by sthalik** [comment]
@kanzaki-kunbr install the 32-bit version for VC++ 2015.

**Comment by sthalik** [comment]
@kanzaki-kunbr the simplest solution is this: https://www.techpowerup.com/download/visual-c-redistributable-runtime-package-all-in-one/

**Comment by kanzaki-kunbr** [comment]
![unknown](https://user-images.githubusercontent.com/104150366/164489927-eb4e9c84-e5b0-4cdb-b8b1-90fab87269e2.png)

it worked thanks for the help

**Comment by jorgelfrias89** [comment]
Hola por alguna razón que desconozco cuando ejecuto el opentrack no me sale en la lista el neuralnet, estuve viendo arriba y tengo las librerías de c++ todas instaladas tanto en 32bit como en 64bit, uso SO Windows 10

**Comment by Aniubik** [comment]
Also having same problem. Already tried everything and still not working. I have al Visuals installed and still nothing. This is showing up in cmd: DEBUG [F:\dev\opentrack\api/plugin-support.hpp:106]: library "neuralnet" failed: "Cannot load library C:\\Program Files (x86)\\opentrack\\modules\\\\opentrack-tracker-neuralnet.dll: A dynamic link library (DLL) initialization routine failed." I have no programing skills, can someone help me please?

**Comment by sthalik** [comment]
It's a Windows 7 issue that now affects way more trackers and I'm looking into it right now.

**Comment by sthalik** [comment]
I switched to `-MD` and guess what? it works

**Comment by sthalik** [comment]
Now's the time you test the build: https://ananke.misaki.pl/opentrack/opentrack-test-20240212_01.7z

**Comment by Tguy3100** [comment]
I seem to be experiencing the same issue neuralnet does not show up in the list however I do have all the redistributables installed Im running windows 10  

DEBUG [F:\dev\opentrack\api/plugin-support.hpp:106]: library "neuralnet" failed: "Cannot load library C:\\Users\\thatg\\OneDrive\\Desktop\\Open track\\install\\modules\\\\opentrack-tracker-neuralnet.dll: A dynamic link library (DLL) initialization routine failed."

not sure why onedrive is involved but I could sure use some help

**Comment by stewi2** [comment]
I'm on Windows 11 and am having the same issue using the latest build. Perhaps this is because I'm running on ARM? Will try to build the DLL myself for lack of a better idea.

**Comment by sthalik** [comment]
@DaWelter can you make a temporary build with tracker/nn that works? Thanks.

**Comment by CaptainCorgi269** [comment]
also couldnt seem to find the neuralnet tracker in windows 11 arm 64 just to mention.  Am not a programmer so im not so sure, i will fiddle around to find a solution.

**Comment by sthalik** [comment]
@Leko20001 https://github.com/opentrack/opentrack/issues/1394#issuecomment-1938864870

**Comment by Leko20001** [comment]
It doesn't work on me

**Comment by Leko20001** [comment]
Also mine is Windows 10

**Comment by sthalik** [comment]
You can use 2023.1.0.

**Comment by Leko20001** [comment]
I have tried to run on 2023.1.0 and I cant even run it
